bigquery_etl/docs/derived_datasets/generate_derived_dataset_docs.py

```diff
@@ -56,6 +57,11 @@ def generate_derived_dataset_docs(out_dir, project_dir):
                 if VIEW_FILE in files:
                     source_urls["View"] = f"{SOURCE_URL}/{root}/{VIEW_FILE}"
 
+                    view_file = os.path.join(root, VIEW_FILE)
+                    referenced = DryRun(view_file).get_referenced_tables()
+                    if referenced:
+                        print('Referenced tables: ', referenced)
+
```

Comment:
@acmiyaguchi So I’m working on getting referenced tables for the user facing views, but there must be something wrong with my (very naive) approach because this prints 0 referenced table for all view files. What am I missing?

Reply:
@wlach Could you help also? Thanks!

Reply:
Are there any print statements logged during the process? You may want to set logging or a trace in the dryrun code to debug:

https://github.com/mozilla/bigquery-etl/blob/1dbb06ce9233c59b9c52a93c36d95f882a56494f/bigquery_etl/dryrun.py#L216-L224

I'd recommend either:

```python
print(self.dry_run_result)

# alternatively, to get into the debugger
import pdb; pdb.set_trace()
```

Reply:
Yes there are print statements logged, but it only prints empty lists for all views:

<img width="1097" alt="Screen Shot 2021-02-05 at 12 17 59 PM" src="https://user-images.githubusercontent.com/28797553/107084749-4d121a80-67ac-11eb-8df7-29ca307913cb.png">

Thanks for the debugging tip, I'll keep investigating.

Reply:
It'd be more informative to look at the dry-run response directly, it may give insight into why the field is missing. I can imagine permission issues being one possible error.

Reply:
Yep it's permission issue. That's so strange I already set up GCP and have access to BigQUery

Reply:
I'm not sure, the dry-run endpoint may be publicly accessible because I don't see an [oidc token](https://cloud.google.com/scheduler/docs/http-target-auth) in the request header. Do you mind sharing a full response with the auth issue?

@scholtzan do you know how the cloud function is configured, and whether the behavior here is expected?

Reply:
The cloud function should be publicly accessible, so no authentication should be required

Reply:
Full error log:

```
{

    'valid':False,
    'errors':[
        {
            'code':403,
            'errors':[
                {
                    'message':'Access Denied: Dataset moz-fx-data-shared-prod:messaging_system: User does not have bigquery.tables.create permission for dataset moz-fx-data-shared-prod:messaging_system.',
                    'domain':'global',
                    'reason':'accessDenied'
                }
            ],
            'response':{
                'headers':{
                    'alt-svc':'h3-29=":443"; ma=2592000,
                    h3-T051=":443"; ma=2592000,
                    h3-Q050=":443"; ma=2592000,
                    h3-Q046=":443"; ma=2592000,
                    h3-Q043=":443"; ma=2592000,
                    quic=":443"; ma=2592000; v="46,
                    43"', 'cache-control':'private',
                    'content-encoding':'gzip',
                    'content-type':'application/json; charset=UTF-8',
                    'date':'Fri,
                    05 Feb 2021 21:10:05 GMT',
                    'server':'ESF',
                    'transfer-encoding':'chunked',
                    'vary':'Origin,
                    X-Origin, Referer', 'x-content-type-options':'nosniff',
                    'x-frame-options':'SAMEORIGIN',
                    'x-xss-protection':'0'
                }
            },
            'message':'Access Denied: Dataset moz-fx-data-shared-prod:messaging_system: User does not have bigquery.tables.create permission for dataset moz-fx-data-shared-prod:messaging_system.'
        }
    ]

}
SyntaxError: JSON.parse: expected property name or '}' at line 1 column 2 of the JSON data
```

Reply:
Oh, right. You need to check if the view is part of the `SKIP` list and skip them since the dry run will fail for them.

Reply:
That would explain a part of the problem, but it's unclear to me why the ones with valid permissions would be returning empty references.

Reply:
Maybe I'm not understanding this correctly but isn't the SKIP check already done here?

https://github.com/mozilla/bigquery-etl/blob/1dbb06ce9233c59b9c52a93c36d95f882a56494f/bigquery_etl/dryrun.py#L211-L215

Reply:
Looks like there is a `return []` missing. Currently, it is just printing out `...Ignoring dryrun results for` and then moves  on to dry running the file anyway

Reply:
The referenced code is from the `get_referenced_tables()` in `dryrun.py`, which already includes `return []`. 
https://github.com/mozilla/bigquery-etl/blob/1dbb06ce9233c59b9c52a93c36d95f882a56494f/bigquery_etl/dryrun.py#L209-L224
Since I imported the function I assumed that the SKIP check is already included, and I think the views that needed to be skipped were skipped, for example:

<img width="970" alt="Screen Shot 2021-02-05 at 12 17 59 PM" src="https://user-images.githubusercontent.com/28797553/107093943-07a91980-67bb-11eb-9cfd-e53cc44ad334.png">

The error that I got is: `Access Denied: Dataset moz-fx-data-shared-prod:messaging_system: User does not have bigquery.tables.create permission for dataset moz-fx-data-shared-prod:messaging_system.` 

If it's not too long an answer, could you help explain what `bigquery.tables.create permission` is?

Reply:
For creating new views, e.g. via `CREATE OR REPLACE VIEW` (which is what all our view definitions look like), the service account running the SQL needs `bigquery.tables.create` permissions. It looks like view definitions are a special case that is handled in the dry run script: https://github.com/mozilla/bigquery-etl/blob/350291bd1a9608d6659f6cbd4c9123f4a366434f/bigquery_etl/dryrun.py#L244
Dry running view definitions always returns an error due to the missing permissions, but if this error is returned dry run will pretend the dry run was successful.

Reply:
In order for you to be able to get referenced tables, you could retrieve the view query and then dry run that one. There is some documentation here on how to access view metadata including the query: https://cloud.google.com/bigquery/docs/view-metadata

Reply:
That makes total sense to me now! Thank you @scholtzan!